Masonry Alignment and Spacing align-tracks and justify-tracks

stwrt opened the pull request:
#### 7da5d95cea57de320f5eebdf30e3cadacb3d7186
<pre>
Masonry Alignment and Spacing align-tracks and justify-tracks
<a href="https://bugs.webkit.org/show_bug.cgi?id=248835">https://bugs.webkit.org/show_bug.cgi?id=248835</a>
rdar://102782713

Reviewed by Tim Nguyen.

Add initial support for align-tracks and justify-tracks in Masonry layout.
These new properties resemble align-content and justify-content, with the addition of being able
to have multiple items so you can specify for each track.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator AlignTracks const):
(WebCore::CSSPrimitiveValue::operator JustifyTracks const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAlignTracks):
(WebCore::CSSPropertyParserHelpers::consumeJustifyTracks):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::alignTracks const):
(WebCore::RenderStyle::justifyTracks const):
(WebCore::RenderStyle::setAlignTracks):
(WebCore::RenderStyle::setJustifyTracks):
(WebCore::RenderStyle::initialAlignTracks):
(WebCore::RenderStyle::initialJustifyTracks):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::StyleGridData):
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::StyleGridData::operator== const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAlignTracks):
(WebCore::Style::BuilderConverter::convertJustifyTracks):

Canonical link: <a href="https://commits.webkit.org/258544@main">https://commits.webkit.org/258544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bfb19d2ec1bd2c3e26af055f4ac38e7d0a8beaf2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102301 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111606 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106279 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12430 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2354 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94642 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109315 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108083 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92783 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/24280 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4952 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25687 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5091 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2144 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11119 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45189 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5868 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6843 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3688,6 +3688,7 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
         case CSSPropertyWebkitTextZoom:
         case CSSPropertyAdditiveSymbols:
         case CSSPropertyAlignmentBaseline:
+        case CSSPropertyAlignTracks:
```
Please file a bug to add animation support for this property and reference it here with a `FIXME`.

graouts replied:
That said, I expect this single line would do:

`new DiscretePropertyWrapper<AlignTracks>(CSSPropertyAlignTracks, &RenderStyle::alignTracks, &RenderStyle::setAlignTracks)`

… and make sure there is WPT coverage for this, if it doesn't exist.

stwrt replied:
@graouts I did try adding that line, and was running into compile errors on line 3477 was running into some compile issues. Will need to look into that more.

nt1m replied:
@stwrt Does `new DiscretePropertyWrapper<const Vector<StyleContentAlignmentData>&>(...)` work?

stwrt replied:
Still seeing build errors

nt1m replied:
@stwrt which ones are you seeing?

nt1m replied:
I'm still curious what the failures are.

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3741,6 +3742,7 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
         case CSSPropertyInsetInline:
         case CSSPropertyInsetInlineEnd:
         case CSSPropertyInsetInlineStart:
+        case CSSPropertyJustifyTracks:
```
Please file a bug to add animation support for this property and reference it here with a `FIXME`.

graouts replied:
That said, I expect this single line would do:

`new DiscretePropertyWrapper<JustifyTracks>(CSSPropertyJustifyTracks, &RenderStyle::justifyTracks, &RenderStyle::setJustifyTracks)`

… and make sure there is WPT coverage for this, if it doesn't exist.

stwrt commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2908,6 +2908,14 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
             return cssValuePool.createIdentifierValue(CSSValueAuto);
         return currentColorOrValidColor(style, style.accentColor());
     }
+    case CSSPropertyAlignTracks: {
```
?

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2908,6 +2908,14 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
     }
+    case CSSPropertyAlignTracks: {
+        auto list = CSSValueList::createCommaSeparated();
+        return list;
```
Why is this returning an empty list?

stwrt replied:
It is not implemented yet

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7265,6 +7265,42 @@ template<CSSPropertyID property> RefPtr<CSSValue> consumeBackgroundSize(CSSParse
+        if (value)
+            parsedValues->append(value.releaseNonNull());
+        else
+            return nullptr;
```
This doesn't seem correct when I look at the MDN syntax. `align-tracks: safe` isn't valid, but this function will consider it valid.

Ditto for `align-tracks: first`. 

From what I'm seeing the syntax is identical to `align-content`, except for the comma separated bit. So you could call whatever `align-content` is using inside the loop.

stwrt replied:
I took these values from the CSSProperties.json. I'll take another pass through  to check these before I ask for a full review.

stwrt replied:
Just to clarify the align-content in CSSProperties.json.

stwrt commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7265,6 +7265,42 @@ template<CSSPropertyID property> RefPtr<CSSValue> consumeBackgroundSize(CSSParse
+        if (range.atEnd())
+            break;
+
+        auto value = consumeIdent<CSSValueNormal, CSSValueFlexStart, CSSValueFlexEnd, CSSValueCenter, CSSValueSpaceBetween, CSSValueSpaceAround, CSSValueSpaceEvenly, CSSValueStretch, CSSValueFirst, CSSValueLast, CSSValueBaseline, CSSValueUnsafe, CSSValueSafe, CSSValueStart, CSSValueEnd>(range);
```
Need to fix

nt1m reviewed (changes requested):
As discussed on Slack, this should use `consumeContentDistributionOverflowPosition` and co.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1303,6 +1305,36 @@ inline GridAutoFlow BuilderConverter::convertGridAutoFlow(BuilderState&, const C
+    Vector<StyleContentAlignmentData> alignTracks;
+    alignTracks.reserveInitialCapacity(valueList.length());
+
+    for (auto val : valueList) {
```
nit: no abbreviations in webkit coding style

`for (auto value : list)`

nt1m commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -1856,6 +1861,8 @@ class RenderStyle {
 
     static GridAutoFlow initialGridAutoFlow() { return AutoFlowRow; }
+    static Vector<StyleContentAlignmentData> initialAlignTracks() { return Vector<StyleContentAlignmentData>(); }
+    static Vector<StyleContentAlignmentData> initialJustifyTracks() { return Vector<StyleContentAlignmentData>(); }
```
I think you want:

`return { initialContentAlignment() };` to match the spec.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7265,6 +7265,46 @@ template<CSSPropertyID property> RefPtr<CSSValue> consumeBackgroundSize(CSSParse
+        if (range.atEnd())
+            break;
+
+        // justify-content property does not allow the <baseline-position> values.
```
```suggestion
        // justify-tracks property does not allow the <baseline-position> values.
```

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1303,6 +1305,36 @@ inline GridAutoFlow BuilderConverter::convertGridAutoFlow(BuilderState&, const C
+    return alignTracks;
+}
+
+inline Vector<StyleContentAlignmentData> BuilderConverter::convertJustifyTracks(BuilderState& builder, const CSSValue& value)
```
This function is identical to the previous one, can we rename the previous one to `convertContentAlignmentDataList` and set the converter in CSSProperties.json respectively?

nt1m commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -1421,13 +1421,49 @@ span
+// start
+// center
+// start
+// end
```
I'd prefer changing the comments for align-content/justify-content/etc. rather than adding all of these commented out values.

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2908,6 +2908,24 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+        auto list = CSSValueList::createCommaSeparated();
+        auto alignTracks = style.alignTracks();
+
+        for (auto x : alignTracks)
```
```suggestion
        for (auto alignment : alignments)
```

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2908,6 +2908,24 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+
+        return list;
+    }
+    case CSSPropertyJustifyTracks: {
```
I'd combine this with the previous case and use a ternary to pick between alignTracks and justifyTracks

nt1m reviewed (changes requested):
Looks a lot better, I have some nits, but my main concern is the lack of tests for parsing. You can add some in imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/parsing

It's important to add some to make sure we accept all the possible combinations needed, and that we reject invalid values correctly.

You can probably copy and expand:
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/align-content-valid.html
*  LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/align-content-invalid.html 
* LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/justify-content-valid.html
*  LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/justify-content-invalid.html

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/align-tracks-computed.html:
```diff
@@ -0,0 +1,37 @@
+test_computed_value("align-tracks", "unsafe end");
+test_computed_value("align-tracks", "safe flex-start");
+
+test_computed_value("align-tracks", "normal, normal")
```
The [shortest serialization rule](https://drafts.csswg.org/cssom/#serializing-css-values) says that: `normal, normal` serializes to `normal`, and since the computed value matches the specified value. This should be:

> test_computed_value("align-tracks", "normal, normal", "normal")

Similarly, `flex-end, start, start, start` will serialize to `flex-end, start` (spec says if there are less values than the number of tracks, the last one fills the rest)

You'll probably want to add some code in ComputedStyleExtractor + StyleProperties to fix that, but feel free to leave the test failing for the initial iteration and file a new bug for this.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/align-tracks-invalid-expected.txt:
```diff
@@ -0,0 +1,21 @@
+PASS e.style['align-tracks'] = "first" should not set the property value
+PASS e.style['align-tracks'] = "flex-start flex-end" should not set the property value
+PASS e.style['align-tracks'] = "last" should not set the property value
+FAIL e.style['align-tracks'] = "left" should not set the property value assert_equals: expected "" but got "left"
```
Why do we successfully parse `left` or `right` ? They should only be successfully parsed for `justify-tracks`, not `align-tracks`.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/align-tracks-invalid.html:
```diff
@@ -0,0 +1,35 @@
+</head>
+<body>
+<script>
+test_invalid_value("align-tracks", "auto");
```
Would be nice to see more list-related tests, especially a testcase with a valid value everywhere, but the first, middle, last one being invalid.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/align-tracks-valid.html:
```diff
@@ -0,0 +1,39 @@
+</head>
+<body>
+<script>
+test_valid_value("align-tracks", "normal");
```
ditto, list tests here would be nice, including serialization tests:
```
test_valid_value("align-tracks", "flex-start, last baseline");
test_valid_value("align-tracks", "normal, normal", "normal");
test_valid_value("align-tracks", "start, flex-end, flex-end, flex-end, flex-end", "start, flex-end");
etc.
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/justify-tracks-computed.html:
```diff
@@ -0,0 +1,35 @@
+<body>
+<div id="target"></div>
+<script>
+test_computed_value("justify-tracks", "normal");
```
ditto here.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/justify-tracks-invalid.html:
```diff
@@ -0,0 +1,36 @@
+</head>
+<body>
+<script>
+test_invalid_value("justify-tracks", "auto");
```
and here

nt1m replied:
@stwrt This still needs list tests

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/justify-tracks-valid.html:
```diff
@@ -0,0 +1,36 @@
+</head>
+<body>
+<script>
+test_valid_value("justify-tracks", "normal");
```
and here.

nt1m commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -675,7 +675,7 @@ block-axis
 multiple
 
-// CSS_PROP_ALIGN_CONTENT
+// CSS_PROP_ALIGN_CONTENT / CSS_PROP_ALIGN_TRACKS
```
we should just change this to:
```suggestion
// align-content / align-tracks
```

This probably followed an older enum name.

nt1m commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -693,7 +693,7 @@ first-baseline
 // stretch
 
-// CSS_PROP_JUSTIFY_CONTENT
+// CSS_PROP_JUSTIFY_CONTENT / CSS_PROP_JUSTIFY_TRACKS
```
ditto here

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -7265,6 +7265,46 @@ template<CSSPropertyID property> RefPtr<CSSValue> consumeBackgroundSize(CSSParse
+        if (range.atEnd())
+            break;
+
+        auto value = consumeContentDistributionOverflowPosition(range, isContentPositionOrLeftOrRightKeyword);
```
```suggestion
        auto value = consumeContentDistributionOverflowPosition(range, isContentPositionKeyword);
```

seems to be correct according to the syntax and your test

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-align/parsing/align-tracks-invalid.html:
```diff
@@ -0,0 +1,38 @@
+test_invalid_value("align-tracks", "unsafe right");
+test_invalid_value("align-tracks", "unsafe");
+
+test_invalid_value("align-tracks", "auto, left");
+test_invalid_value("align-tracks", "foo, auto");
```
I think ones like `start, left, end`/`left, start, end`/`start, end, left` would be worth adding too

nt1m reviewed (approved):
Can you file follow bugs for serialization and animation support?

stwrt committed:
Masonry Alignment and Spacing align-tracks and justify-tracks
https://bugs.webkit.org/show_bug.cgi?id=248835
rdar://102782713

Reviewed by Tim Nguyen.

Add initial support for align-tracks and justify-tracks in Masonry layout.
These new properties resemble align-content and justify-content, with the addition of being able
to have multiple items so you can specify for each track.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator AlignTracks const):
(WebCore::CSSPrimitiveValue::operator JustifyTracks const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAlignTracks):
(WebCore::CSSPropertyParserHelpers::consumeJustifyTracks):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::alignTracks const):
(WebCore::RenderStyle::justifyTracks const):
(WebCore::RenderStyle::setAlignTracks):
(WebCore::RenderStyle::setJustifyTracks):
(WebCore::RenderStyle::initialAlignTracks):
(WebCore::RenderStyle::initialJustifyTracks):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
(WebCore::StyleGridData::StyleGridData):
* Source/WebCore/rendering/style/StyleGridData.h:
(WebCore::StyleGridData::operator== const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertAlignTracks):
(WebCore::Style::BuilderConverter::convertJustifyTracks):

Canonical link: https://commits.webkit.org/258544@main
